pkg-manager/plugin-commands-installation/test/catalog.ts

```diff
@@ -0,0 +1,324 @@
+import { type ProjectManifest } from '@pnpm/types'
+import { readProjects } from '@pnpm/filter-workspace-packages'
+import { preparePackages } from '@pnpm/prepare'
+import { install } from '@pnpm/plugin-commands-installation'
+import { DEFAULT_OPTS } from './utils'
+import path from 'path'
+import readYamlFile from 'read-yaml-file'
+import writeYamlFile from 'write-yaml-file'
+
+/**
+ * A utility to make writing catalog tests easier by reducing boilerplate.
+ */
+class CatalogTestsController {
+  private readonly workspaceDir: string
+
+  constructor (readonly pkgs: Array<{ location: string, package: ProjectManifest }>) {
+    preparePackages(pkgs)
+    this.workspaceDir = process.cwd()
+  }
+
+  async writeWorkspaceYml (data: unknown) {
+    await writeYamlFile(path.join(this.workspaceDir, 'pnpm-workspace.yaml'), data)
+  }
+
+  async install (opts?: { frozenLockfile?: boolean, filter?: readonly string[] }) {
+    const { allProjects, selectedProjectsGraph } = await readProjects(this.workspaceDir, opts?.filter?.map(parentDir => ({ parentDir })) ?? [])
+    await install.handler({
+      ...DEFAULT_OPTS,
+      allProjects,
+      dir: this.workspaceDir,
+      lockfileDir: this.workspaceDir,
+      recursive: true,
+      selectedProjectsGraph,
+      workspaceDir: this.workspaceDir,
+      useBetaCatalogsFeat: true,
+
+      rawLocalConfig: {
+        ...DEFAULT_OPTS.rawLocalConfig,
+        // Many tests check if the pnpm-lock.yaml is updated after a
+        // package.json or pnpm-workspace.yaml change. Setting frozen-lockfile
+        // to false is necessary for these tests to pass on CI since this
+        // defaults to true on CI.
+        'frozen-lockfile': opts?.frozenLockfile ?? true,
+      },
+    })
+  }
+
+  async lockfile (): Promise<Lockfile> {
+    return readYamlFile(path.join(this.workspaceDir, WANTED_LOCKFILE))
+  }
+
+  async updateProjectManifest (location: string, manifest: ProjectManifest): Promise<void> {
+    const { selectedProjectsGraph } = await readProjects(this.workspaceDir, [{ parentDir: location }])
+    await selectedProjectsGraph[path.join(this.workspaceDir, location)].package.writeProjectManifest(manifest)
+  }
+}
+
+test('installing with "catalog:" should work', async () => {
+  const ctrl = new CatalogTestsController([
+    {
+      location: 'packages/project1',
+      package: {
+        dependencies: {
+          'is-positive': 'catalog:',
+        },
+      },
+    },
+    // Empty second project to create a multi-package workspace.
+    {
+      location: 'packages/project2',
+      package: {},
+    },
+  ])
+
+  await ctrl.writeWorkspaceYml({
+    packages: ['packages/*'],
+    catalog: { 'is-positive': '1.0.0' },
+  })
+  await ctrl.install()
+  const lockfile = await ctrl.lockfile()
+
+  expect(lockfile.importers['packages/project1']).toEqual({
+    dependencies: {
+      'is-positive': {
+        specifier: 'catalog:',
+        version: '1.0.0',
```

Comment:
if the version is also present in the catalogs object, then why is it duplicated here? I guess it makes sense to duplicated it in case when is-positive has peer dependencies. Probably should have such test when the same catalog is used in two projects and has peers that resolve to different versions.

Reply:
Right, it's the peer dependencies part of this. Will do — I'll add a test for that.